### ISSUE TITLE
MCP-0021: Comonent Iterators

### ISSUE THREAD

**Issue by modelica-trac-importer** [tbeu/ModelicaSpecification#641]
**Reported by otter on 2 Dec 2015 08:34 UTC**
This MCP is provided here:

  https://svn.modelica.org/projects/MCP/MAinternal/MCP-0021_ComponentIterators

It consists of

 * Rational.
 * A Modelica test library with many examples.
 * The proposal has been implemented and verified with Dymola and SimulationX

A concrete proposal for the specification change is not yet added, because first an open issue should be resolved.

It is planned to discuss and finalize this proposal at the next Modelica Design Meeting.

Please, add your comments to this ticket.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1848

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 2 Dec 2015 08:36 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 8 Dec 2015 13:59 UTC**
The MCP-0021 proposal has been updated:

 * For-loop only over instances on the same or a lower level (so no guard is needed, and not a loop over all instances in the same simulation model).
 * Added a new [document](https://svn.modelica.org/projects/MCP/MAinternal/MCP-0021_ComponentIterators/MCP-0021_ComponentIterators_SpecChanges.pdf) containing the proposed changes to the Modelica Specification text.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 13 Sep 2016 10:26 UTC**
The proposal is now at:
https://svn.modelica.org/projects/MCP/public/MCP-0021_ComponentIterators/MCP-0021_ComponentIterators_SpecChanges.pdf

Two issues were found:
* The "this level or below" means that the idea of dragging in an observer model no longer works (would require going up one level more), it is important to get this right to make it useable. Without the restriction at all we get the problem of separating out relevant components to observer. Having the observers as base-classes sort of work with this restriction - but causes a risk of name-conflicts. One level above would need an exception if already at the top of the simulation model - but would risk that we compute cost of everything in model - and then we add this as a component in a larger model and it adds cost of sibling components.
Without any restriction we would get the same problem - one possibility is to have some level-argument.

* Using the instance name (by `IDENT.getInstanceName()`) prevents using the new record constructor syntax for this.
A work-around is to put instance-name inside the components, but that require modifying all the models - which wasn't the intention.

The basic of the record constructor idea is that `Record(v)` is mapped to `Record(r1=v.r1, r2=v.r2, ...)`, so instead of writing `Record(r1=v.r1, r2=v.r2, ...)` we can write `Record(v)`.
But what to write instead of: 
`Record(r1=v.r1, r2=v.r2, n=v.getInstanceName(), ...)`. 
Could we interpret:

```
record Record
  Real r1, r2;
  String n=getInstanceName();
  ...
end Record;
```
and interpret `Record(v)` as `Record(r1=v.r1, r2=v.r2, n=v.getInstanceName(), ...)`? Perhaps this would avoid the "IDENT.getInstanceName()"?
Another possibility is to make "instanceName" a special component that is always "present" - you just have to include it in your record.

One possibility is not solve that part (for now): the names may anyway be different so new record constructor does not work, use a sub-record for the non-name part, or just ignore the name.
\--
Need to resolve the level-issue, and can then put under evaluation.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 3 Nov 2016 13:04 UTC**
One possibility for the level-issue is to have:

For an iterator:
 IDENT in class [ UNSIGNED_INTEGER ] NAME
the NAME must be (using normal lexical lookup) the name of a model, block, connector, record, or operator record specialized class. The loop-variable, IDENT, is resolved to every instance of class NAME - the optional UNSIGNED_INTEGER argument indicates that the search is restricted to the component sub-tree that many levels above the component declaration. If no instance of class NAME is present, the constructed array has zero dimension.
\--
I don't find it optimal - but I believe we can make it work.
The instance name part can (as indicated) be delayed.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Nov 2016 09:10 UTC**
Added description:

Without the level (UNSIGNED_INTEGER) restriction all components will be considered. If we have a mechanical model that uses this to compute the total mass of parts, and makes two instances of this mechanical model both of the instances will have double the mass of the single mechanical model.

The level restriction can be used to avoid that and encapsulate a model to a re-usable sub-system that is not influenced by the outside: both directly to handle just components in the current class ("in class 0 ...") - or as reusable component that considers components on the same level as itself ("in class 1 ...").

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 5 Dec 2016 13:46 UTC**
At the meeting it was proposed to move this one level up, and also to more clearly define "component level" - especially for arrays. That has now been done in r9545.

I believe this is more useful, but we can also discuss the previous variant (which requires having the observer-model as base-class) - or even the level argument as above.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 13 Dec 2016 10:39 UTC**
Text updated in r9564.
Move to "Under Evaluation":
Agree:  7
Against: 0
Rest abstain.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 14 Dec 2016 13:40 UTC**
Updated MCP-documents by changing status to "Under Evaluation" and updating Excel-File https://svn.modelica.org/projects/MCP-summary.xlsx

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 24 Jan 2017 15:56 UTC**
> [...] the constructed array has zero dimension.

This should be "the constructed array has zero length in the corresponding dimension." - it doesn't change the number of dimensions.
> [...] `model`, `block`, `connector`, `record`, or `operator record`, `type` specialized class.

The "or" is in the wrong place, it should be "`model`, `block`, `connector`, `record`, `operator record`, or `type` specialized class."

I'd also prefer that "the NAME must be" is changed to just "NAME must be".

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 25 Jan 2017 13:02 UTC**

In the specification text the sentence
> The loop-variable, IDENT, is resolved to every instance of class NAME that is on the same or a lower level as the parent of the component containing the iterator expression.

has some problems:
- it states "same or lower level" and "parent of the component". One of them is too much to fulfill the example in the proposal below. Either "same" or "parent" should be removed.
- the term "containing" is not well specified, especially if the iterator is used in a modification. Then it is unclear whether the component is the parameter itself or its parent or the component defined by the surrounding class which corresponds to the term scope in the flattening MCP-0019. 
Anyway, the user has to know the result of the instantiation process of the model.

In rationale document several restrictions are mentionened. Some of them are not part of the specification proposal, but if they hold, they should be there, e.g.:
> \* Component iterators can only be used in the specialized classes model and block.
> \* An instance iterator inside an “outer component” is ignored.
> \* Conditionally disabled objects are ignored.

In this list the second point is also unclear. How can only an instance iterator beeing ignored in an outer component. Usually outer components are only "symbolic links" to the corresponding inner component and are completely ignored if they are not inner by themselves.
Also, the iterator should not list outer objects which are not marked also as inner. This should be in the specification.

Another point which is only mentioned in the rational but not in the specification is the handling of arrays of CLASS. Also the example of record instances getting their value from others is not handled.

Unclear is also whether protected elements are listed by the iterator. Probably yes, but is this the intention of protected? If not the result will be wrong.

In general, I find the proposal a littlebit narrow. Because of the nominal resolution of NAME, it only finds components of exact this class. Inheritence and structural subtyping are not recognized.
Therefore it forces the library developer to a specific modelling style, if such iterator should be used. In particular he has to avoid inheritance, which is bad in general and contradicts to an object oriented modelling.
On the other hand, a user of this pattern can not be sure at the first place, whether he finds all existing components or not. He has to inspect the library in detail for that.
This could be improved by adding already existing keywords to the iterator definition, e.g. 
- `IDENT in class extends NAME` for inheritance or 
- `IDENT in class constrainedby NAME` for structural subtyping.
Even then it is nearly impossible to know that.

To summarize: I would propose to not include it in 3.4 but to improve and add it later.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 30 Jan 2017 08:58 UTC**
The vote on this is now completed:
 7 Accept and add to 3.4
 1 Accept without adding to 3.4
 4 Do not accept.

So, Accepted and planned for 3.4, but we should investigate and discuss the feedback at the next meeting.
